### PR TITLE
feat(transport): make vehicle a useful logistical doctype for VRP solvers

### PR DESCRIPTION
**Pull request by blaggacao**
# Context

Consider being a Fleet Manager with a routing optimization problem at hand and multimodal transport.

The current vehicle data entry is no match and too specific to motor vehicles.

# Proposed Solution
- Rename (too specific) "Vehicle" to (more generic) "Transport"
- Add Mode & Propulsion for as anchor values for further customization
- Remove overly vehicle specific field: `last_carbon_check` (-> customize if needed)
- Add Vehicle Routing Problem parameter section (after having implemented VRP in production I know for a fact that these are the metadata values for: cost/distance matrix, capacity dimensions and capability matching)

# Screenshots
image
image
image

### CODE REVIEW

File: erpnext/setup/doctype/vehicle/vehicle.json [hunk 1]
```diff
@@ -597,20 +586,7 @@
...
-   "print_hide": 0,
-   "print_hide_if_no_value": 0,
-   "read_only": 0,
-   "remember_last_selected_value": 0,
-   "report_hide": 0,
-   "reqd": 1,
-   "search_index": 0,
-   "set_only_once": 0,
-   "unique": 0
+   "options": "\nPetrol\nDiesel\nNatural Gas\nElectric\nMuscle"
```
Comment (bosue) [inline erpnext/setup/doctype/vehicle/vehicle.json]: Unsure about adding “muscle”. This seems more funny than relevant.
“Hydrogen” and “Hybrid” is missing, too.
E-Bikes might be considered hybrid, too.
Maybe just add “Other”?

Comment (blaggacao) [reply erpnext/setup/doctype/vehicle/vehicle.json]: Yes, you're right. This doctype was so entrenched on cars that it's still far away from a general understanding of what a vehicle may be. So this became inadvertently a bit funny. 😁

I'll add these suggestions, though.

Comment (bosue) [reply erpnext/setup/doctype/vehicle/vehicle.json]: I agree. There‘s so much more to do until being useful for cars, bikes and other vehicles alike. But your PR will at least remove car-focussed requirements that are currently showstoppers for bikes. 👍

Comment (blaggacao) [reply erpnext/setup/doctype/vehicle/vehicle.json]: https://github.com/frappe/erpnext/pull/37321/commits/e17a6aa9e0e2b8272c68f068e525733c5fd367f8

Comment (blaggacao) [reply erpnext/setup/doctype/vehicle/vehicle.json]: Ok, this now has gotten a little rewiring based on my experiences of running a VRP solver in ERPNext in production now for a couple of months. This is an iteration and if accepted, I may consider to release the VRP app in the future.

Comment (bosue) [reply erpnext/setup/doctype/vehicle/vehicle.json]: Now it's coming together, yes!
I was kind of wondering what it is all about, but with the VRP solver in mind, it completely makes sense. 👍

**Comment by bosue** [comment]
Now that I understand what this is aiming at, there's one aspect causing me a bit of headache:
For a while I've been thinking about how to best couple Vehicle with Asset.

We need all the details such as license plate, make, model, location etc. plus a few more, most importantly the car's VIN, on the Asset doctype which currently doesn't give any specifics.
Renaming Vehicle to Transport will further abstract away the specifics of cars, which may be welcome for VRP, but not so much when considering the Vehicle as an Asset.

Unsure how to solve the dilemma best. I think the word "Vehicle" does actually encompass bikes, e-bikes, scooters etc., so it might be better to keep it, unless trains, busses and other riding opportunities without Asset characteristics are to be included as well?

**Comment by blaggacao** [comment]
Trains, Airplanes, Donkeys may be all part of the equation, yes.

Other than that, I had exactly the same feeling about Assets as @bosue.

Vehicles _was_ also a poor man's asset management (hybrid) doctype and I feel that a tighter integration with asset management may be helpful, but leaning on the ERPNext's asset doctypes, rather than a setup doctype.

I can imagine that an asset is simply linked to a transport (or the other way round?).

For additional transport specific fields, I had considered the possibility to customize entire sections and make them transport type specific for the lack of doctype inheritance for further specialization.

So if it's a car, I thought customizations could add a section "Car" with car-specific values, etc

That was the best I could think of up to now maybe there's a better way?

**Comment by bosue** [comment]
> Trains, Airplanes, Donkeys may be all part of the equation, yes.

For VRP yes, absolutely. In terms of an Asset not so much...

> customize entire sections and make them transport type specific

Not 100% sure if you meant customizing the Asset or - probably - the ~Vehicle~ Transport doctype. But either way, I don't think that's particularly attractive, if viable.

Assets with Vehicle characteristics / details are a major usecase that many, many companies should need in their books. 
Their mileage or health may require Asset Value Adjustements and/or influence the depreciation plan. Their insurance status needs to be integrated with all other Assets' insurance status. Asset Repairs will in turn be relevant to the Vehicle doctype's integration with HRMS. And whatever holds all relevant data may and will be the basis of lots of additional features, such as a VIN checker, GPS tracking, service tracking, Carbon compensation etc.

All of this needs a single place where things are coming together, not several of them and certainly not the customization of a secondary, merely coupled doctype that caters to trains and donkeys as well...

> for the lack of doctype inheritance for further specialization

That's unfortunately a big part of our problem, yes. We'd need Assets that are Vehicles and others that are Machines, maybe even different kinds of machines with all their various specifics, and while adding specific features as customizations may be viable, building those on top of some kind of JS-simulated polymorphism seems way too fragile for the specific usecase and unnecessarily complex for the basic one.

While some Frappe support for full visual integration of coupled doctypes would be nice, in the end only full and even multiple inheritance as in: _Vehicle is an Asset_, and _Vehicle is a Transport_ would give us everything we want, but of course it would also bring a whole dimension of new problems.

First steps though? I'd say let's:

- keep Vehicle and add a Means of Transport (or Transport Option) doctype.
- allow coupling Asset with a Vehicle.
- allow coupling Means of Transport with one or several Vehicles. Finally you may need two vehicles - such as a semi-truck with a semi-trailer - as a single viable transport option. And in the next transport you will have to couple the same semi-truck with another semi-trailer to create a transport option for certain goods.
- consolidate Insurance details in an Asset Insurance or maybe even more general an Insurance doctype that may allow linking one or more Assets.
- consider moving the whole Fleet Management module-def with Driver, Vehicle Log, Vehicle Service Item, and Vehicle Expenses from HRMS either back to ERPNext, or together with Vehicle, Transport (and possibly VRP) to a brandnew Transport app. These are not necessarily tied to employees at all, but certainly not primarily tied to them. And they are currently relatively lightweight, while HRMS may still add employee specific doctypes, reports, procedures.

But again this is a large plan. Within this issue's limits, I'm just proposing to keep Vehicle as is and add a separate "Means of Transport" or "Transport Option" doctype coupled 1:n.
Just "Transport" is quite a bit ambiguous. Don't know if that's a language barrier thing but I would think of it as a single transport from A to B, which may be either a delivery or a pickup of goods or a passenger transport...

**Comment by blaggacao** [comment]
Oh I didn't realize there's more fleet management inside HRMS. That changes the perspective and makes a separate transportation app actually not all that un-attractive.

As to the name 'Transport', it's as per one of the english language dictionaries just the right word for a single thing that moves something ("Transportmittel", which im German can be thing or class alike, not so in english, as it seems). "Mode of Transport", in turn, is a class of things and "Vehicle" is already a subclass of those things: vehicles. A series of legs, in turn, may be called an itinerary or trip, made out of "legs" (edges) and "stops" (nodes).

All financial aspect would probably be best moved over to assets, such as insurance, although with a special permission for "Fleet Managers" to manage those data entry items.

A good balance can't be perfect, because that would make the overall system too inflexible. Some non-integration might be left necessary.

**Comment by blaggacao** [comment]
Unstale

**Comment by rmehta** [comment]
@blaggacao can you publish this as a custom app? I don't think a full scale vehicle management system should part of core ERP.

**Comment by blaggacao** [comment]
@rmehta , yes indeed. It has only very little overlap with the core, except for this Doctype.

I've put this im draft and only would undraft once I find the time to publish it as a separate app as accompanying patch.

**Comment by blaggacao** [comment]
Nope this shouldn't necessarily go stale, it's just draft.

**Pull request closed without merging**